Library/OcBootManagementLib/BootArguments.c

```diff
@@ -360,3 +374,226 @@ OcCheckArgumentFromEnv (
+          if (State == LOAD_OPTIONS_WHITE_SPACE) {
+            DEBUG ((OC_TRACE_PARSE_LOAD_OPTIONS, "OCB: LoadOption no value %u\n", 1));
+          }
+          Name = NULL;
+        }
+        break;
+
+      case LOAD_OPTIONS_VALUE_FIRST:
+        if (c == L'"') {
+          State = LOAD_OPTIONS_QUOTED_VALUE;
+          Value = Pos + 1;
+        } else {
+          State = LOAD_OPTIONS_VALUE;
+          Value = Pos;
+          Retake = TRUE;
+        }
+        break;
+
+      case LOAD_OPTIONS_VALUE:
+      case LOAD_OPTIONS_QUOTED_VALUE:
+        if (c == L'\\') {
+          SHIFT_TOKEN (Pos, Value, 1);
+        } if (
+          (State == LOAD_OPTIONS_VALUE && (IS_SPACE (c) || c == CHAR_NULL)) ||
+          (State == LOAD_OPTIONS_QUOTED_VALUE && c == '"')) {
+          if (Pos - Value == 1) {
+            DEBUG ((OC_TRACE_PARSE_LOAD_OPTIONS, "OCB: LoadOption no value %u\n", 2));
+          } else {
+            *Pos = CHAR_NULL;
+            Option->Value = Value;
+            DEBUG ((OC_TRACE_PARSE_LOAD_OPTIONS, "OCB: LoadOption Value=\"%s\"\n", Value));
+          }
+          Value = NULL;
+          Option = NULL;
+          State = LOAD_OPTIONS_WHITE_SPACE;
+        }
+        break;
+
+      default:
+        ASSERT (FALSE);
+        break;
+    }
+
+    if (Retake) {
+      Retake = FALSE;
+    } else {
+      ++Pos;
+    }
+  } while (c != CHAR_NULL);
+
+  if (State != LOAD_OPTIONS_WHITE_SPACE) {
+    DEBUG ((DEBUG_ERROR, "OCB: Invalid LoadOptions (%u) (%p:%u)\n", State, LoadedImage->LoadOptions, LoadedImage->LoadOptionsSize));
+    OcDynamicArrayFree (ParsedLoadOptions);
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((*ParsedLoadOptions)->Items == NULL) {
+    DEBUG ((DEBUG_WARN, "OCB: Empty LoadOptions\n"));
+    OcDynamicArrayFree (ParsedLoadOptions);
+    return EFI_NOT_FOUND;
+  }
+
+  return EFI_SUCCESS;
+}
+
+BOOLEAN
+OcLoadOptionGetStr (
+  IN     CONST OC_DYNAMIC_ARRAY   *ParsedLoadOptions,
+  IN     CONST CHAR16             *Name,
+     OUT       CHAR16             **Value
+  )
+{
+  UINTN           Index;
+  OC_LOAD_OPTION  *Option;
+
+  ASSERT (Name != NULL);
+  ASSERT (Value != NULL);
+
+  if (ParsedLoadOptions == NULL) {
+    return FALSE;
+  }
+
+  ASSERT (ParsedLoadOptions->Items != NULL);
+
+  for (Index = 0; Index < ParsedLoadOptions->Count; Index++) {
+    Option = OcDynamicArrayItemAt (ParsedLoadOptions, Index);
+    ASSERT (Option->Name != NULL);
+    if (StrCmp (Option->Name, Name) == 0) {
+      *Value = Option->Value;
+      DEBUG ((OC_TRACE_PARSE_LOAD_OPTIONS, "OCB: Using LoadOption \"%s\"=\"%s\"\n", Name, *Value));
+      return TRUE;
+    }
+  }
+
+  DEBUG ((OC_TRACE_PARSE_LOAD_OPTIONS, "OCB: No LoadOption \"%s\"\n", Name));
+  return FALSE;
+}
+
+BOOLEAN
+OcHasLoadOption (
+  IN     CONST OC_DYNAMIC_ARRAY   *ParsedLoadOptions,
+  IN     CONST CHAR16             *Name
+  )
+{
+  CHAR16      *Value;
+
+  return OcLoadOptionGetStr (ParsedLoadOptions, Name, &Value);
```

Comment:
Use explicit != NULL checking.

Reply:
The intention was the NULL checking was already done inside the inner call - should be outside as well?

Reply:
No, I mean return value checking. Do not rely on NULL being implicitly FALSE.

Reply:
Ah, OcLoadOptionGetStr returns BOOLEAN.

Reply:
Because it is possible for an option to be present, with a NULL string value.

Reply:
Ah, never mind. Thanks.